fix: deselect toolbar button when dragging the cursor out of the toolbar (fixes #2088)

trevinhofmann opened the pull request:
Previously, a toolbar button would get stuck down if the user clicks it and drags the mouse off the toolbar before releasing. This fixes that by listening for the `mouseLeave` event and deselecting the toolbar button when the mouse leaves the toolbar.

**Demo:**

https://github.com/user-attachments/assets/627779aa-3ffc-4d06-a5cf-d21faf47dd30

raineorshine reviewed (changes requested):
Looks good!

There is one case where there is a behavior mismatch. If you drag back onto the original toolbar button and release the mouse button, the button surprisingly activates. So if you're going to visually deactivate the button on mouseleave, then the behavior on mouseup also needs to be deactivated for consistency.

The other approach is to visually deactivate the toolbar button on mouseup, wherever the cursor is, instead of on mouseleave. Then the current behavior would make sense. I'm fine with either approach.

Let me know if that makes sense. Thanks!

trevinhofmann commented:
@raineorshine

**Please note:** I have rebased this on top of the `trevinhofmann:trevin/tweak-toolbar-button-animation` branch because there were minor conflicts and dependencies on changes in that branch. Please review the PR for that branch (https://github.com/cybersemics/em/pull/2214) before completing the review of this one. The following commits are taken from that other branch:
* [fix: ensure that the full toolbar button animation plays](https://github.com/cybersemics/em/pull/2210/commits/d88379b59581b4bfe6afec450217ea179804ae74)
* [refactor: hoist TOOLBAR_PRESS_ANIMATION_DURATION into constants.ts](https://github.com/cybersemics/em/pull/2210/commits/60c641e5cde817ac6bbc56e1255e266e0b35a4eb)
* [refactor: use well-named intermediate variables to clarify toolbar button deselection calculation](https://github.com/cybersemics/em/pull/2210/commits/c1f6b8a3b60e8295c41240eb6b7c77855dcc730c) 
* [refactor: remove onTapCancel from the toolbar buttons](https://github.com/cybersemics/em/pull/2210/commits/95acb5ceda43d481ffd62d3fb606544e3a20220f)

trevinhofmann commented on src/components/Toolbar.tsx:
```diff
@@ -74,6 +74,11 @@ const Toolbar: FC<ToolbarProps> = ({ customize, onSelect, selected }) => {
+  /** Handles mouse leave event. */
+  const onMouseLeave = useCallback(() => {
+    setPressingToolbarId(null)
+  }, [])
```
That makes sense! Implemented in this commit: 9bbf5d5c90b4fbaca141029d09128d3c22bae301

trevinhofmann commented on src/components/Toolbar.tsx:
```diff
@@ -87,6 +92,15 @@ const Toolbar: FC<ToolbarProps> = ({ customize, onSelect, selected }) => {
   }, [updateArrows])
 
+  // deselect button when moving the cursor out of the toolbar
+  const toolbarContainer = toolbarContainerRef.current
```
Thank you. I was experiencing some finicky conflicting behavior with the linter and the behavior of the dependencies in the hook, but making that change in this commit worked as expected this time: bbfaabfb347a7778a3bd78ca2e45bb4b72181c72

trevinhofmann commented:
@raineorshine -- Yes, the issue of dragging the cursor back over the button again and then releasing made sense and was easy to reproduce. I fixed that behavior in this commit: 6a91d1d31ad6e85d648dd293f540ec0b8daf1656

While testing that case, I also realized that we are only deselecting the button when dragging outside of the entire toolbar. That also leads to strange functionality such as pressing down on "Bold", dragging over to "Italic", and then pressing up. I have fixed that behavior as well by moving the `onMouseLeave` listener to the button rather than the toolbar, in this commit: 5e20484cdb6ded7aa9e24fa3ea31591a77ae0cf5

Here is an updated demo showing the improved functionality:

https://github.com/user-attachments/assets/9f71b056-2788-476b-ae7f-0b6e81b81aaa

trevinhofmann commented:
The commit 6a91d1d31ad6e85d648dd293f540ec0b8daf1656 caused test failures in `src/components/__tests__/ColorPicker.ts`. They failed because the test utility only simulated a single `click` event, but the toolbar buttons now expect a click to consist of both the mouse/press down and mouse/press up.

I implemented 7ffe4bbdba44d0bb665f2b6b06e3b752220a49fc to adjust the test utility to simulate both events separately. @raineorshine, please let me know if that is reasonable. At the moment, that utility is only used in the ColorPicker test file, so I don't believe there would be other unintended side effects.

raineorshine commented on src/components/ToolbarButton.tsx:
```diff
@@ -140,10 +142,20 @@ const ToolbarButtonComponent: FC<DraggableToolbarButtonProps> = ({
     [buttonError, colors, dropToRemove, fontSize, isButtonActive, isButtonExecutable, isDragging, longPress.isPressed],
   )
 
+  // deselect button when moving the cursor off the button
+  useEffect(() => {
+    const toolbarButton = toolbarButtonRef.current
+    if (toolbarButton && onMouseLeave) {
+      toolbarButton.addEventListener('mouseleave', onMouseLeave)
+      return () => toolbarButton.removeEventListener('mouseleave', onMouseLeave)
+    }
+  }, [toolbarButtonRef, onMouseLeave])
```
Just thought of this, but would it work to just put `onMouseLeave={onMouseLeave}` on the div?

trevinhofmann replied:
That absolutely does work, and this hook is quite silly in hindsight 😅

Refactored: e7a9697ea0fd89408ac29abece6b799381da8c16

raineorshine commented on src/test-helpers/click.ts:
```diff
@@ -1,11 +1,13 @@
 const click = async (selector: string) => {
   const el = document.querySelector(selector)!
   await act(async () => {
-    userEvent.click(el)
+    fireEvent.mouseDown(el)
+  })
+  await act(async () => {
+    fireEvent.mouseUp(el)
   })
```
Does this work in a single `act`? That would be more performant.

trevinhofmann replied:
I did test that, but it only works as expected in a separate act.

raineorshine reviewed (changes requested):
> **Please note:** I have rebased this on top of the `trevinhofmann:trevin/tweak-toolbar-button-animation` branch because there were minor conflicts and dependencies on changes in that branch. Please review the PR for that branch (#2214) before completing the review of this one. The following commits are taken from that other branch:

I just merged #2144. I think because I squashed commits this is giving merge conflicts now, but they should be easy to fix.

> While testing that case, I also realized that we are only deselecting the button when dragging outside of the entire toolbar. That also leads to strange functionality such as pressing down on "Bold", dragging over to "Italic", and then pressing up. I have fixed that behavior as well by moving the `onMouseLeave` listener to the button rather than the toolbar, in this commit: [5e20484](https://github.com/cybersemics/em/commit/5e20484cdb6ded7aa9e24fa3ea31591a77ae0cf5)

I actually noticed that behavior, and was going to file it as a separate bug. Thank you!

trevinhofmann commented:
Thanks! I reset this branch to `react-trevin` and then cherry-picked the new commits on top of that. It should be good to go now, @raineorshine.

raineorshine reviewed (approved):
Perfect, thanks!
